dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java

```diff
@@ -134,18 +140,44 @@ public boolean moveNext() {
         leftMatchingRowsIterator = null;
         while (right.moveNext()) {
             int rightHash = hashBuilderForRight.apply(right.currentElement());
-            Collection<Object[]> leftMatchingRows = buffer.get(rightHash);
-            if (leftMatchingRows != null) {
-                leftMatchingRowsIterator = leftMatchingRows.iterator();
+            Object[] leftRowFromBuffer = buffer.get(rightHash);
+            if (leftRowFromBuffer != null) {
+                leftRow.cells(leftRowFromBuffer);
+                combiner.setLeft((L) leftRow);
                 combiner.setRight(right.currentElement());
-                if (findMatchingRows()) {
+                if (joinCondition.test(combiner.currentElement())) {
                     return true;
                 }
+            } else if (!bufferWithHashCollisions.isEmpty()) {
+                List<Object[]> leftMatchingRows = bufferWithHashCollisions.get(rightHash);
+                if (leftMatchingRows != null) {
+                    leftMatchingRowsIterator = leftMatchingRows.iterator();
+                    combiner.setRight(right.currentElement());
+                    if (findMatchingRows()) {
+                        return true;
+                    }
+                }
             }
         }
         return false;
     }
 
+    private void addToBuffers(Object[] currentRow, int hash) {
+        Object[] existingRow = buffer.get(hash);
+        List<Object[]> existingRows = bufferWithHashCollisions.get(hash);
+        if (existingRows == null && existingRow == null) {
+            buffer.put(hash, currentRow);
```

Comment:
I don't think this put is necessary as well, if there's nothing in the map under that hash, a subsequent get will still return null

Reply:
This is the building phase of the hashmap. If the element doesn't exist yet in the buffers we put it first to the simple buffer were we hope there are no hash collisions, duplicate values.

Reply:
ah you're right it's the _currentRow_ we're adding. great